### PR TITLE
[JENKINS-56575] Restart API returns 200 when request comes from REST client

### PR DESCRIPTION
**Pull request by ydli-ai**
❗️ Closed in favor of #4030

----

See [JENKINS-56575](https://issues.jenkins-ci.org/browse/JENKINS-56575).

When user using web browser to *restart* Jenkins, the operation will be checked in `_restart.jelly` which makes the POST request is referred from */restart* page. When a rest client directly send POST request to restart Jenkins, the request has no `Referer`. So we know the request come from browser or rest client by checking whether `Referer` is null.

This bug should be fixed because 503 is not a reasonable response to REST API, which would make rest client perform unnecessary retry.
 

### Proposed changelog entries

* Entry 1: Change HTTP response code of restart API to 200 when request is not from web browser

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

### Desired reviewers

@jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -4192,6 +4192,13 @@ public void doRestart(StaplerRequest req, StaplerResponse rsp) throws IOExceptio
             return;
         }
 
+        //Request from rest client
+        if (req.getReferer() == null || req.getMethod().equals("POST")) {
```
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java]: NPE when `req == null`

**Review by daniel-beck — DISMISSED** [review]

**Comment by varyvol** [comment]
@wadatka does this have any security-related implication?

**Comment by daniel-beck** [comment]
I think this ping was for @Wadeck

**Review by Wadeck — COMMENTED** [review]
Thank you for the contribution. I totally agree the situation is a bit weird for REST client in such case :)

The use of `referer` header as a distinction factor between regular HTTP client and pure REST client is misleading. You can encounter troubles if a user has privacy oriented configuration, like [Firefox sendRefererHeader](http://kb.mozillazine.org/Network.http.sendRefererHeader).
If you read the linked page you will see that there is a warning that this option can break compatibility with some websites. Generally that's the case when the application is using the referer as the way to counter CSRF (wrong way but that's another topic).

I can only advice you to use `Accept` header with `application/json` or an additional parameter in the request to distinguish between regular and rest clients (preference for the accept option).
Both proposals are easy to apply on the rest client, could it be simple `curl`, `Postman` or anything else.

Please look at #3972 where I proposed the `Accept` approach.

**Comment by daniel-beck** [comment]
Superseded by #3972 based on positive author feedback for the alternative approach.

**Pull request closed without merging**